internal/provisioner/utility_group.go

```diff
@@ -0,0 +1,69 @@
+package provisioner
+
+import (
+	"github.com/mattermost/mattermost-cloud/internal/tools/aws"
+	"github.com/mattermost/mattermost-cloud/internal/tools/kops"
+	"github.com/mattermost/mattermost-cloud/model"
+)
+
+// ClusterUtility is a third party service that runs one per cluster and must be managed at cluster creation/deletion time. It is responsible for management of its own lifetime, given a handle pointing to the correct cluster.
+type ClusterUtility interface {
+	Create(group UtilityGroup) error
+	Upgrade(group UtilityGroup) error
+	Destroy(group UtilityGroup) error
+}
+
+// UtilityGroup holds the metadata needed to manage a specific UtilityGroup, and therefore uniquely identifies one, and can be thought of as a handle to the real group of utilities running inside of the cluster
+type UtilityGroup struct {
+	cluster     *model.Cluster
+	provisioner *KopsProvisioner
+	awsClient   aws.AWS
+	kops        *kops.Cmd
+}
```

Comment:
A few things on this:

1. I like the move to an interface here! It really gives us a flexible way to work on things.
2. We should pass a logging object into the utilities and not use the provisioner logger directly. The reason for this is that we add new fields to our loggers in different processes so it would be nice to pass that specific logger in to keep those fields. See the kops CreateCluster flow to get an example of what I mean.
3. UtilityGroup is slightly confusing to me. I would think this would refer to a group of Utilities, but it's actually more like tools for the utilities to do their work. Maybe `UtilityTools` instead? Before you think too much about that though, see my next comment.
4. I think it would do well to change the logic here a bit. Instead of passing in a bunch of tools/things that may be needed by a utility, I think we may want to have a `New()` function for each utility type. The idea here is each utility will init with only the resources it needs and those functions will not be part of the utility interface. That way the `New()` function can also validate the passed in resources. i.e. it can check for things like the various pointers being `nil`. Then the `ClusterUtility` interface will change to something like:
```
type ClusterUtility interface {
	Create() error
	Upgrade() error
	Destroy() error
}
```

Reply:
> We should pass a logging object into the utilities and not use the provisioner logger directly. The reason for this is that we add new fields to our loggers in different processes so it would be nice to pass that specific logger in to keep those fields. See the kops CreateCluster flow to get an example of what I mean.

I _think_ I've done what you're asking here -- used `WithField` to create a new logging object to pass around whenever needing a new logger instead of just reusing the provisioner object. Please have a look and let me know if I have or have not done the right thing.